core/include/usBundleContext.h

```diff
@@ -691,7 +736,18 @@ class US_Core_EXPORT BundleContext
   friend class BundlePrivate;
   template<class S> friend struct ServiceHolder;
 
+  // allow templated code to use the internal logger
+  template<class S, class TTT, class R> friend class BundleAbstractTracked;
+  template<class S, class T> friend class ServiceTracker;
+  template<class S, class TTT> friend class ServiceTrackerPrivate;
+  template<class S, class TTT> friend class TrackedService;
+
   BundleContext(BundlePrivate* bundle);
+
+  // Not for use by clients of the Framework.
+  // Provides access to the Framework's log sink to allow templated code
+  // to log diagnostic information. 
+  std::shared_ptr<LogSink> GetLogSink();
```

Comment:
You are right, the template code makes this really hard if not 
impossible. I agree that the private method in BundleContext is a good 
compromise.


On 07/27/2016 07:58 PM, Jeff DiClemente wrote:
>
> In core/include/usBundleContext.h 
> <https://github.com/CppMicroServices/CppMicroServices/pull/96#discussion_r72489954>:
>
> >    BundleContext(BundlePrivate* bundle);
> > +
> > +  // Not for use by clients of the Framework.
> > +  // Provides access to the Framework's log sink to allow templated code
> > +  // to log diagnostic information.
> > +  std::shared_ptr<LogSink> GetLogSink();
>
> I started working on this and am realizing that this may be more 
> trouble than its worth.
>
> I can't see a way to push |GetLogSink()| into |BundleContextPrivate| 
> without exposing some other private header/function. Even if I was to 
> export |GetPrivate()| from the library (like what is done with 
> |MakeBundleContext()| in |usGetBundleContext.h|) clients can then 
> access the |BundleContextPrivate| object by doing 
> |GetPrivate(MakeBundleContext(...))|.
>
> I'm not sure this is worth the effort given our timeline for 3.0 right 
> now.
> Is there a better solution which doesn't expose private implementation 
> classes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/CppMicroServices/CppMicroServices/pull/96/files/f233328c1a7c45f4e0e623708c41be235d302864#r72489954>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AANXpQ-A_CRponMXseX8K8pfgCRUu6-7ks5qZ5wvgaJpZM4IcUhJ>.
>